docs/source/examples/qlora_finetune.rst

```diff
@@ -0,0 +1,283 @@
+converted tensors to the CPU. This offloading is to avoid peaking memory by maintaining an entire bf16/fp32 copy of the ``state_dict``
+on GPU, which could lead to potential OOMs during checkpoint save, even if memory is appropriately managed during
+training.
+
+
+
+Putting it all together: QLoRA finetune
+-----------------------------------------
+
+Putting it all together, we can now finetune a model using TorchTune's `LoRA recipe <https://github.com/pytorch/torchtune/blob/48626d19d2108f92c749411fbd5f0ff140023a25/recipes/lora_finetune.py>`_,
+with a `<QLoRA configuration https://github.com/pytorch/torchtune/blob/main/recipes/configs/llama2/7B_qlora_single_device.yaml>`_.
+
+Make sure that you have first downloaded the Llama2 weights and tokenizer by following :ref:`these instructions<download_llama_label>`.
+You can then run the following command to perform a QLoRA finetune of Llama2-7B using the Alpaca dataset on a single GPU.
+
+.. code-block:: bash
+
+    tune run lora_finetune_single_device --config recipes/configs/llama2/7B_qlora_single_device.yaml
+
+.. note::
+    Make sure to correctly point to the location of your Llama2 weights and tokenizer. This can be done
+    either by adding :code:`checkpointer.checkpoint_files=[my_model_checkpoint_path] tokenizer_checkpoint=my_tokenizer_checkpoint_path`
+    or by directly modifying the :code:`7B_qlora_single_device.yaml` file. See our :ref:`config_tutorial_label`
+    for more details on how you can easily clone and modify TorchTune configs.
+
+By default, this run should log peak memory stats at model initialization time and every 100
+iterations during training. Let's understand the memory savings enabled by QLoRA on top of LoRA training. LoRA training
+can be run as follows:
+
+.. code-block:: bash
+
+    tune run lora_finetune_single_device --config recipes/configs/llama2/7B_lora_single_device.yaml
+
+You should see the memory usage printed out during model initialization:
+
+.. code-block:: python
+
+  Memory Stats after model init::
+  GPU peak memory allocation: 13.96 GB
+  GPU peak memory reserved: 13.98 GB
+  GPU peak memory active: 13.96 GB
+
+
+As well as during training:
+
+.. code-block:: python
+
+  Memory Stats::
+  GPU peak memory allocation: 14.40 GB
+  GPU peak memory reserved: 15.57 GB
+  GPU peak memory active: 14.40 GB
+
+
+Comparing to the memory usage during model initialization for QLoRA, we see about a 35% decrease in peak memory reserved:
+
+.. code-block:: python
+
+  Memory Stats after model init::
+  GPU peak memory allocation: 7.36 GB
+  GPU peak memory reserved: 9.13 GB
+  GPU peak memory active: 7.36 GB
+
+As well as a 40% decrease in peak memory reserved during training:
+
+.. code-block:: python
+
+  Memory Stats::
+  GPU peak memory allocation: 5.54 GB
+  GPU peak memory reserved: 9.29 GB
+  GPU peak memory active: 5.54 GB
```

Comment:
I'm a bit torn here.. while I think it's nice to print the raw output that someone will see in the logs, personally I would just do it one time, e.g. "you will see something like this" and then put it in a table (QLoRA, LoRA) x (peak init memory, peak training memory). Then it's a bit more digestible

Reply:
Added table